core/arch/arm/plat-hikey/spi_test.c

```diff
@@ -28,73 +28,236 @@
 #include <drivers/pl022_spi.h>
 #include <drivers/pl061_gpio.h>
 #include <hikey_peripherals.h>
+#include <io.h>
+#include <kernel/tee_time.h>
 #include <stdint.h>
 #include <trace.h>
+#include <util.h>
 
-/*
- * spi_init() must be run before calling this function.
- *
- * This runs a loopback test by default, so the SPI module will just
- * receive what is transmitted, i.e. 0x01, 0x80, 0x00.
- *
- * In non-loopback mode, the transmitted value will elicit a readback of
- * the measured value from the ADC chip on the Linksprite 96Boards
- * Mezzanine card [1], which can be connected to either a sliding
- * rheostat [2] or photoresistor [3].
- *
- * [1] http://linksprite.com/wiki/index.php5?title=Linker_Mezzanine_card_for_96board
- * [2] http://learn.linksprite.com/96-board/sliding-rheostat
- * [3] http://learn.linksprite.com/96-board/photoresistor
- */
-void spi_test(void)
+#define PL022_STAT	0x00C
+#define PL022_STAT_BSY	SHIFT_U32(1, 4)
+
+static void spi_cs_callback(enum gpio_level value)
 {
 	struct pl061_data platform_pl061_data;
-	struct pl022_data platform_pl022_data;
 	vaddr_t gpio6_base = nsec_periph_base(GPIO6_BASE);
 	vaddr_t spi_base = nsec_periph_base(SPI_BASE);
+
+	pl061_init(&platform_pl061_data);
+	pl061_register(gpio6_base, 6);
+	pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
+	platform_pl061_data.chip.ops->set_interrupt(GPIO6_2,
+						GPIO_INTERRUPT_DISABLE);
+	platform_pl061_data.chip.ops->set_direction(GPIO6_2, GPIO_DIR_OUT);
```

Comment:
Is it ok to do below?

```
	static bool is_init = false;

	if (!is_init) {
		pl061_init(&platform_pl061_data);
		pl061_register(gpio6_base, 6);
		pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
		platform_pl061_data.chip.ops->set_interrupt(GPIO6_2,
							GPIO_INTERRUPT_DISABLE);
		platform_pl061_data.chip.ops->set_direction(GPIO6_2, GPIO_DIR_OUT);
		is_init = true;
	}
```

Otherwise, if there's a separate init routine, we'll have to pass a pointer to `pl061_data` to the callback function, which will make it dependent on pl061 driver, which kinds of defeat the purpose of the callback, which is to make it flexible for the caller to implement a CS callback of their design.